src/platform/web/ui/session/room/timeline/TextMessageView.js

```diff
@@ -32,22 +32,86 @@ export class TextMessageView extends BaseMessageView {
     }
 }
 
+function renderList(listBlock) {
+    const items = listBlock.items.map(item => tag.li({}, renderParts(item)));
+    const start = listBlock.startOffset;
+    if (start) {
+        return tag.ol({ start }, items);
+    } else {
+        return tag.ul({}, items);
+    }
+}
+
+function renderImage(imagePart) {
+    const attributes = Object.assign(
+        { src: imagePart.src },
+        imagePart.width && { width: imagePart.width },
+        imagePart.height && { height: imagePart.height },
+        imagePart.alt && { alt: imagePart.alt },
+        imagePart.title && { title: imagePart.title }
+    );
+    return tag.img(attributes, []);
+}
+
+function renderPill(pillPart) {
+    // The classes and structure are borrowed from avatar.js;
+    // We don't call renderStaticAvatar because that would require
+    // an intermediate object that has getAvatarUrl etc.
+    const classes = `avatar size-12 usercolor${pillPart.avatarColorNumber}`;
+    const avatar = tag.div({class: classes}, text(pillPart.avatarInitials));
+    const children = renderParts(pillPart.children);
+    children.unshift(avatar);
+    return tag.a({ class: "pill", href: pillPart.href }, children);
+}
+
+function renderTable(tablePart) {
+    const children = [];
+    if (tablePart.head) {
+        const headers = tablePart.head
+            .map(cell => tag.th({}, renderParts(cell)));
+        children.push(tag.thead({}, tag.tr({}, headers)))
+    }
+    const rows = [];
+    for (const row of tablePart.body) {
+        const data = row.map(cell => tag.td({}, renderParts(cell)));
+        rows.push(tag.tr({}, data));
+    }
+    children.push(tag.tbody({}, rows));
+    return tag.table({}, children);
+}
+
 /**
  * Map from part to function that outputs DOM for the part
  */
 const formatFunction = {
+    header: headerBlock => tag["h" + Math.min(6,headerBlock.level)]({}, renderParts(headerBlock.inlines)),
+    codeblock: codeBlock => tag.pre({}, tag.code({}, text(codeBlock.text))),
+    table: tableBlock => renderTable(tableBlock),
+    emph: emphPart => tag.em({}, renderParts(emphPart.inlines)),
+    code: codePart => tag.code({}, text(codePart.text)),
     text: textPart => text(textPart.text),
-    link: linkPart => tag.a({ href: linkPart.url, target: "_blank", rel: "noopener" }, [linkPart.text]),
+    link: linkPart => tag.a({ href: linkPart.url, target: "_blank", rel: "noopener" }, renderParts(linkPart.inlines)),
+    pill: renderPill,
+    format: formatPart => tag[formatPart.format]({}, renderParts(formatPart.children)),
+    list: renderList,
+    image: renderImage,
     newline: () => tag.br()
 };
 
+function renderPart(part) {
+    const f = formatFunction[part.type];
+    return f(part);
+}
+
+function renderParts(parts) {
+    return Array.from(parts, renderPart);
+}
+
 class BodyView extends StaticView {
     render(t, messageBody) {
         const container = t.span();
```

Comment:
As things stand, all the block elements end up inside the span. That doesn't seem correct (we just recently talked about something like this in the intermediate representation), and the subsequent timestamp looks a bit weird.

Reply:
Have fixed this by using mapSideEffect, if we ever decide to put the time elsewhere we can get rid of this hack-ish thing.